pkg/controllers/manila/manila.go

```diff
@@ -0,0 +1,218 @@
+package manila
+
+import (
+	"context"
+	"reflect"
+	"time"
+
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack/sharedfilesystems/v2/sharetypes"
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/csi-driver-manila-operator/pkg/util"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/resource/resourcemerge"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	storagev1 "k8s.io/api/storage/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	storagelisters "k8s.io/client-go/listers/storage/v1"
+	"k8s.io/klog/v2"
+)
+
+// This Controller watches OpenStack and:
+// 1) Installs Manila CSI drivers (Manila itself, NFS) once
+//    it detects that there is Manlina present (by running provided
+//    manilaOperatorSet).
+// 2) Creates StorageClass for each share type provided by Manila.
+// 3) If there is no Manila in the OpenStack where the cluster runs,
+//    it marks the operator with condition Disabled=true.
+//
+// Note that the CSI driver(s) are not un-installed when Manila becomes
+// missing or it stops providing shares of given type - Manila bight be
+// under (short?) maintenance / reconfiguration.
+// Similarly, StorageClasses are not deleted when a share type disappears
+// from Manila.
+type Controller struct {
+	operatorClient     v1helpers.OperatorClient
+	kubeClient         kubernetes.Interface
+	secretLister       corelisters.SecretLister
+	openStackClient    *openStackClient
+	storageClassLister storagelisters.StorageClassLister
+	// Controllers to start when Manila is detected
+	csiControllers     []Runnable
+	controllersRunning bool
+	eventRecorder      events.Recorder
+}
+
+type Runnable interface {
+	Run(ctx context.Context, workers int)
+}
+
+const (
+	// Minimal interval between controller resyncs. The controller will detect
+	// new share types in Manila and create StorageClasses for them at least
+	// once per this interval.
+	resyncInterval = 1 * time.Minute
```

Comment:
Do we need to resync every minute?

Reply:
When someone creates a new share type in OpenStack, this operator should create StorageClass for it. Question is what's the right frequency.

Reply:
IIUC the optimal resync interval for this type of storage backend (manila, EFS) will vary from cluster to cluster. For instance, in some clusters the number of share types will seldom change, while in others new share types will be added and removed quite often.

Having said that, whatever we hardcode here might not be the ideal for some use cases.

IMO the best way to approach this is to define a standard value, say 20 minutes, and allow the cluster admin to customize it through the CR. If we choose this approach, we could tackle the customization part in another PR. 

Thoughts?

Reply:
I'm updating it to 20 min, however, I am not sure how important it is to have it customizable.